design/kata-api-design.md

```diff
@@ -0,0 +1,112 @@
+### Sandbox Management API
+|Name|Description|
+|---|---|
+|CreateSandbox(SandboxConfig)| Create, start a sandbox and return the sandbox structure.|
+|FetchSandbox(ID)| Connect to an existing sandbox and return the sandbox structure.|
+|ListSandbox()| List all existing standboxes with status. |
+
+### Sandbox Operation API
+|Name|Description|
+|---|---|
+|sandbox.Pause()| Pause the sandbox.|
+|sandbox.Resume()| Resume the paused sandbox.|
+|sandbox.Delete()| Destroy the sandbox and remove all persistent metadata.|
+|sandbox.Status()| Get the status of the sandbox and containers.|
+|sandbox.Monitor()| Return a context handler for caller to monitor sandbox callbacks such as error termination.|
+|sandbox.CreateContainer()| Create new container in the sandbox.|
+|sandbox.DeleteContainer()| Delete a container from the sandbox.|
+|sandbox.StartContainer()| Start a container in the sandbox.|
+|sandbox.StatusContainer()| Get the status of a container in the sandbox.|
+|sandbox.EnterContainer()| Run a new process in a container.|
+|sandbox.WaitProcess()| Wait on a process.|
+### Sandbox Hotplug API
+|Name|Description|
+|---|---|
+|sandbox.AddStorage()| Add new storage to the sandbox.|
+|sandbox.AddNetwork()| Add new nic to the sandbox.|
+|sandbox.ListNetwork()| List all nics and their configurations in the sandbox.|
+|sandbox.UpdateNetwork()| Update the configuration of an existing network interface.|
+|sandbox.UpdateRoute()| Update the sandbox route table, e.g. for portmapping support.|
+
+### Sandbox Relay API
+|Name|Description|
+|---|---|
+|sandbox.WinsizeProcess(containerID, processID, Height, Width)|Relay TTY resize request to a process.|
+|sandbox.SignalProcess(containerID, processID, signalID, signalALL)| Relay a signal to a process or all processes in a container.|
+|sandbox.IOStream(containerID, processID)| Relay a process stdio. Return stdin/stdout/stderr pipes to the process stdin/stdout/stderr streams.|
+
+## Plugin framework for external proprietary Kata runtime extensions
+### Hypervisor plugin
+
+TBD.
+### Metadata storage plugin
+The metadata storage plugin controls where sandbox metadata is saved.
+All metadata storage plugins must implement following API:
+
+|Name|Description|
+|---|---|
+|storage.Save(key, value)| Save a record.|
+|storage.Load(key)| Load a record.|
+|storage.Delete(key)| Delete a record.|
+
+Built-in implementations include:
+   -  Filesystem storage
+   -  LevelDB storage
+
+### VM Factory plugin
+The VM factory plugin controls how a sandbox factory creates new VMs.
+All VM factory plugins must implement following API:
+
+|Name|Description|
+|---|---|
+|VMFactory.NewVM(HypervisorConfig)|Create a new VM based on `HypervisorConfig`.|
+
+Built-in implementations include:
+
+|Name|Description|
+|---|---|
+|CreateNew()| Create brand new VM based on HypervisorConfig|
+|CreateFromTemplate()| Create new VM from template.|
+|CreateFromCache()| Create new VM from VM caches.|
+
+### Sandbox Creation Plugin Workflow
```

Comment:
The diagram looks fine, but should be updated with hypervisor implementations and not `plugins`. Also the title should be:
```
Sandbox Creation Workflow
```

Reply:
@sboeuf the workflow is mainly to explain when VM Factory plugin and metadata plugin APIs are called.

Reply:
Since `CreateSandbox()` is performing multiple operations, internally this should be using some sort of transaction manager to ensure an atomic outcome:

- all operations completed successfully.
- an error occurred, so all *successful* operations are undone (to return the system to the state it was at before `CreateSandbox()` was called.

In summary: should we state that all APIs are atomic in operation?

Reply:
@jodh-intel yes, we would want the sandbox operation APIs atomic to roll back any in-progress state if anything goes wrong.